src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -3068,6 +3073,155 @@ impl Coordinator {
         Ok(ExecuteResponse::AlteredObject(ObjectType::Secret))
     }
 
+    #[instrument]
+    pub(super) async fn sequence_alter_sink_prepare(
+        &mut self,
+        ctx: ExecuteContext,
+        plan: plan::AlterSinkPlan,
+    ) {
+        // 1. Put a read hold on the new relation
+        let id_bundle = crate::CollectionIdBundle {
+            storage_ids: BTreeSet::from_iter([plan.sink.from]),
+            compute_ids: BTreeMap::new(),
+        };
+        let read_hold = self.acquire_read_holds(&id_bundle);
+
+        let Some(read_ts) = read_hold.least_valid_read().into_option() else {
+            ctx.retire(Err(AdapterError::UnreadableSinkCollection));
+            return;
+        };
+
+        let otel_ctx = OpenTelemetryContext::obtain();
+
+        let plan_validity = PlanValidity {
+            transient_revision: self.catalog().transient_revision(),
+            dependency_ids: BTreeSet::from_iter([plan.sink.from]),
+            cluster_id: Some(plan.in_cluster),
+            replica_id: None,
+            role_metadata: ctx.session().role_metadata().clone(),
+        };
+
+        // Re-resolve items in the altered statement
+        // Parse statement.
+        let create_sink_stmt = match mz_sql::parse::parse(&plan.sink.create_sql)
+            .expect("invalid create sink sql")
+            .into_element()
+            .ast
+        {
+            Statement::CreateSink(stmt) => stmt,
+            _ => unreachable!("invalid statment kind for sink"),
+        };
+        let catalog = self.catalog().for_system_session();
+        let (_, resolved_ids) = match mz_sql::names::resolve(&catalog, create_sink_stmt) {
+            Ok(ok) => ok,
+            Err(e) => {
+                ctx.retire(Err(AdapterError::internal("ALTER SINK", e)));
+                return;
+            }
+        };
+
+        // Now we must wait for the sink to make enough progress such that there is overlap between
+        // the new `from` collection's read hold and the sink's write frontier.
+        self.controller.install_storage_watch_set(
+            BTreeSet::from_iter([plan.id]),
+            read_ts,
+            Box::new(WatchSetResponse::AlterSinkReady(AlterSinkReadyContext {
+                ctx,
+                otel_ctx,
+                plan,
+                plan_validity,
```

Comment:
I don't see where this field is read and re-validated after it gets back to the main coord loop.

Reply:
This happens here https://github.com/MaterializeInc/materialize/pull/27379/files#diff-d1bdc6aec56cf00c650cbc1c9f811b1dc96d3e0a46190aed7033c40ea0591738R447